src/probe/probe.c

```diff
@@ -15,42 +24,551 @@
 
 int probe_deinit(void)
 {
+	uint32_t i;
+	int err;
+
+	tracev_probe("probe_deinit()");
+
+	if (!_probe) {
+		trace_probe_error("probe_deinit() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	/* check for attached injection probe DMAs */
+	for (i = 0; i < CONFIG_PROBE_DMA_MAX; i++) {
+		if (_probe->inject_dma[i].stream_tag != PROBE_DMA_INVALID) {
+			trace_probe_error("probe_deinit() error: Cannot deinitialize with injection DMAs attached.");
+			return -EINVAL;
+		}
+	}
+
+	/* check for connected probe points */
+	for (i = 0; i < CONFIG_PROBE_POINTS_MAX; i++) {
+		if (_probe->probe_points[i].stream_tag != PROBE_POINT_INVALID) {
+			trace_probe_error("probe_deinit() error: Cannot deinitialize with probe points connected.");
+			return -EINVAL;
+		}
+	}
+
+	if (_probe->ext_dma.stream_tag != PROBE_DMA_INVALID) {
+		tracev_probe("probe_deinit() Freeing extraction DMA.");
+		err = probe_dma_deinit(&_probe->ext_dma);
+		if (err < 0)
+			return err;
+	}
+
+	rfree(_probe);
+	_probe = NULL;
+
 	return 0;
 }
 
 int probe_dma_add(uint32_t count, struct probe_dma *probe_dma)
 {
+	uint32_t i;
+	uint32_t j;
+	uint32_t stream_tag;
+	uint32_t first_free;
+	int err;
+
+	tracev_probe("probe_dma_add() count = %u", count);
+
+	if (!_probe) {
+		trace_probe_error("probe_dma_add() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	for (i = 0; i < count; i++) {
+		tracev_probe("\tprobe_dma[%u] stream_tag = %u, dma_buffer_size = %u",
+			     i, probe_dma[i].stream_tag,
+			     probe_dma[i].dma_buffer_size);
+
+		first_free = CONFIG_PROBE_DMA_MAX;
+
+		/* looking for first free dma slot */
+		for (j = 0; j < CONFIG_PROBE_DMA_MAX; j++) {
+			stream_tag = _probe->inject_dma[j].stream_tag;
+
+			if (stream_tag == PROBE_DMA_INVALID) {
+				if (first_free == CONFIG_PROBE_DMA_MAX)
+					first_free = j;
+
+				continue;
+			}
+
+			if (stream_tag == probe_dma[i].stream_tag) {
+				trace_probe_error("probe_dma_add() error: Probe DMA %u already attached.",
+						  stream_tag);
+				return -EINVAL;
+			}
+		}
+
+		if (first_free == CONFIG_PROBE_DMA_MAX) {
+			trace_probe_error("probe_dma_add() error: Exceeded maximum number of DMAs attached = " META_QUOTE(CONFIG_PROBE_DMA_MAX));
+			return -EINVAL;
+		}
+
+		_probe->inject_dma[first_free].stream_tag =
+			probe_dma[i].stream_tag;
+		_probe->inject_dma[first_free].dma_buffer_size =
+			probe_dma[i].dma_buffer_size;
+
+		err = probe_dma_init(&_probe->inject_dma[first_free]);
+		if (err < 0) {
+			trace_probe_error("probe_dma_add() error: probe_dma_init() failed");
+			_probe->inject_dma[first_free].stream_tag =
+				PROBE_DMA_INVALID;
+			return err;
+		}
+	}
+
 	return 0;
 }
 
 int probe_dma_info(struct sof_ipc_probe_info_params *data, uint32_t max_size)
 {
+	uint32_t i = 0;
+	uint32_t j = 0;
+
+	tracev_probe("probe_dma_info()");
+
+	if (!_probe) {
+		trace_probe_error("probe_dma_info() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	data->rhdr.hdr.size = sizeof(*data);
+
+	/* iterate over inject DMAs */
```

Comment:
why iterate, what are we looking for, please make the comments describe what the block is doing e.g. "search all injection DMAs for X in order to do Y"

Reply:
This applies to lots of the other comments too.